mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java

```diff
@@ -669,17 +665,19 @@ private void trimFlags(Flags flags, MailboxSession session) {
         }, MailboxPathLocker.LockType.Write);
     }
 
-    protected MessageMetaData appendMessageToStore(final MailboxMessage message, final List<MessageAttachment> messageAttachments, MailboxSession session) throws MailboxException {
-        final MessageMapper messageMapper = mapperFactory.getMessageMapper(session);
+    private MessageMetaData appendMessageToStore(Date internalDate, int size, int bodyStartOctet, SharedInputStream content, Flags flags, PropertyBuilder propertyBuilder, MailboxSession session) throws MailboxException {
+        MessageMapper messageMapper = mapperFactory.getMessageMapper(session);
+        MessageId messageId = messageIdFactory.generate();
 
         return mapperFactory.getMessageMapper(session).execute(() -> {
-            storeAttachment(message, messageAttachments, session);
+            List<MessageAttachment> attachments = storeAttachments(messageId, content, session);
+            MailboxMessage message = createMessage(internalDate, size, bodyStartOctet, content, flags, propertyBuilder, attachments);
```

Comment:
I really think that it's not the responsibility of storeAttachments to generate the MessageAttachement ids, It makes everything much more complex than it should be.

Reply:
It's a common pattern we have to have the storage layer generating ids

I don't really alternatives regarding this...

Reply:
We had the same problem with blob-store recently. We decided to give the responsibility to the client. It's IMO a very important design decision.
In the case of this PR it will make some things simpler.